src/containers/Signup/Signup.tsx

```diff
@@ -24,65 +41,72 @@ const INITIAL_FORM_FIELDS: IFormFields = {
   confirmationCode: ''
 };
 
-const Signup: FC<IAppProps & RouteComponentProps> = props => {
+const Signup: FC<RouteComponentProps> = ({ history }) => {
+  const { push: historyPush } = history;
+  const dispatch = useDispatch();
+
+  const signingUp = useSelector(selectAuthSigningUp);
+  const signedUp = useSelector(selectAuthSignedUp);
+  const signUpError = useSelector(selectAuthSignUpError);
+
+  const activating = useSelector(selectAuthActivating);
+  const activated = useSelector(selectAuthActivated);
+  const activateError = useSelector(selectAuthActivateError);
+
+  const signingIn = useSelector(selectAuthSigningIn);
+  const signedIn = useSelector(selectAuthSignedIn);
+  const signInError = useSelector(selectAuthSignInError);
+
   const [fields, handleFieldChange] = useFormFields<IFormFields>(
     INITIAL_FORM_FIELDS
   );
-  const [newUser, setNewUser] = useState<ISignUpResult | null>(null);
-  const [isLoading, setIsLoading] = useState(false);
+
+  const { email, password } = fields;
+
+  useEffect(() => {
+    if (signUpError) alert(signUpError);
+    if (activateError) alert(activateError);
+    if (signInError) alert(signInError);
+  }, [signUpError, activateError, signInError]);
+
+  useEffect(() => {
+    if (activated) dispatch(signInAuthAction(email, password));
+    if (signedIn) {
+      dispatch(checkedAuthAction(true));
+      dispatch(resetAuthInnerStateAction('signin'));
+      dispatch(resetAuthInnerStateAction('signup'));
+      dispatch(resetAuthInnerStateAction('activate'));
+      historyPush('/');
+    }
+  }, [activated, signedIn, historyPush, dispatch, email, password]);
```

Comment:
Why do all these functions to be in a single useEffect?

Reply:
Maybe my question should be, why not:
```
  useEffect(() => {
    if (activated) dispatch(signInAuthAction(email, password));
  }, [activated, email, password]);

  useEffect(() => {
    if (signedIn) {
      dispatch(checkedAuthAction(true));
      dispatch(resetAuthInnerStateAction('signin'));
      dispatch(resetAuthInnerStateAction('signup'));
      dispatch(resetAuthInnerStateAction('activate'));
      historyPush('/');
    }
  }, [signedIn, historyPush, dispatch]);
```

If they needed to be in the same useEffect, it was not obvious why or what business logic it is trying to achieve for me. If I were to revisit the code few months later I would be lost.

Btw you don't need to watch `dispatch` probably as well right?

Reply:
Apologize, I didn't quite understand, could suggest your solution other wise.

Regarding the `dispatch` react is showing a warning on start. Saying missing dependency `dispatch` on useEffect

Reply:
> Regarding the dispatch react is showing a warning on start. Saying missing dependency dispatch on useEffect
Okay no helping there.

Reply:
oh right right.
I just wanted to clear/re-initialize the internal state of auth.

```
authState: {
  signin: {
    signingIn: boolean;
    signedIn: boolean;
  },
  activate: {
    activating: boolean;
    activated: boolean;
  },
  signup: {
    signingUp: boolean;
    signedUp: boolean;
  },
  check: {
    checking: boolean;
    checked: boolean;
  }
}
```

Now you might be thinking, why can't we just clear or re-initialize the whole `authState`.
Due to the `check` internal state. I need the checked property wether to redirect to `/signin` page or continue on the authenticated routes i.e `/home`, `/note.`

I'm clearing up the other `auth` internal state due to when the user decided to `/signout` and go to `/signin` or `/signout`  if we leave those to their state without clearing/re-initializing, there will be bugs across the loading buttons it will end up showing `loading` and on `/signup`, `activation-form` will render instead of the `signup-form` since `signedUp` property is already `true`.

Business logic:

User signs up -> success -> render `activation form` -> activates -> success -> automated sign in -> success -> re-initialize state for `signUp`, `activate` and `signIn` -> re-direct `/`